File: sdk/java/src/main/java/jp/kengos/shojiku/TemplateRoot.java
package jp.kengos.shojiku;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.function.Predicate;
import java.util.regex.Pattern;

/**
 * Resolving a template NAME to the sources behind it.
 *
 * <p>A name is an identifier, never a path. A bundle format will take this lookup over later, so
 * nothing outside this class may assume a directory is how names resolve — callers ask for {@code
 * "receipt_ja"} and get sources back.
 *
 * <p>THE REJECTION RULES ARE THE UNION ACROSS PLATFORMS, NOT THE HOST'S. Windows is a first-class
 * target, so a backslash is a separator, {@code C:name} is drive-relative, {@code \\host\share} is
 * a UNC path and CON/NUL are reserved devices — every one of them refused on EVERY platform. A
 * template name that is valid on one machine is valid on all of them, which is the only way the
 * same application deploys to both.
 */
final class TemplateRoot {

  static final String TEMPLATE_FILE = "templates.yml";
  static final String DEFINITIONS_FILE = "definitions.yml";

  /**
   * Reserved DOS device names.
   *
   * <p>Windows resolves these no matter what directory you are in and no matter what extension you
   * append.
   */
  private static final Set<String> DEVICES = buildDevices();

  // A name is ONE segment. Refusing both separators outright subsumes traversal,
  // absolute paths and nested lookups in a single rule — the simplest thing six
  // other SDKs can mirror without drifting.
  private static final Pattern SEPARATORS = Pattern.compile("[/\\\\]");
  private static final Pattern DRIVE_RELATIVE = Pattern.compile("\\A[A-Za-z]:");
  private static final Pattern TRAILING_DOTS_AND_SPACES = Pattern.compile("[.\\s]+\\z");

  /** Each rule, and what a caller is told when it fires. */
  private static final List<Rule> RULES =
      List.of(
          new Rule(
              TemplateRoot::isSeparator,
              "a name is one segment, so `/` and `\\` are never part of it "
                  + "(which is also what makes `..` traversal impossible)"),
          new Rule(TemplateRoot::isControl, "it contains a control character"),
          new Rule(
              TemplateRoot::isDriveRelative,
              "it is drive-relative, which Windows resolves against that drive's current directory"),
          new Rule(TemplateRoot::isDevice, "it is a reserved device name on Windows"));

  private final String path;

  TemplateRoot(String path) {
    this.path = path;
  }

  String path() {
    return path;
  }

  /** Resolve {@code name}, or throw {@link RejectedException} naming why it will not. */
  Sources resolve(String name) {
    reject(name);
    Path real = contained(Paths.get(path, name));
    return new Sources(
        read(real.resolve(TEMPLATE_FILE)),
        optional(real.resolve(DEFINITIONS_FILE)),
        real.toString());
  }

  private static Set<String> buildDevices() {
    Set<String> devices = new java.util.HashSet<>(Set.of("CON", "PRN", "AUX", "NUL"));
    for (int n = 1; n <= 9; n++) {
      devices.add("COM" + n);
      devices.add("LPT" + n);
    }
    return Set.copyOf(devices);
  }

  private static boolean isSeparator(String name) {
    return SEPARATORS.matcher(name).find();
  }

  private static boolean isControl(String name) {
    return name.chars().anyMatch(Character::isISOControl);
  }

  private static boolean isDriveRelative(String name) {
    return DRIVE_RELATIVE.matcher(name).find();
  }

  /**
   * Trailing dots and spaces are STRIPPED by Windows before it resolves a name.
   *
   * <p>So {@code CON.} and {@code "CON "} are the CON device just as {@code CON} is. Without that
   * strip they slip past this rule and are refused later, by containment — still refused, but with
   * a message about a missing template rather than about a reserved name.
   */
  private static boolean isDevice(String name) {
    String stem = TRAILING_DOTS_AND_SPACES.matcher(name.split("\\.", -1)[0]).replaceAll("");
    return DEVICES.contains(stem.toUpperCase(Locale.ROOT));
  }

  private static void reject(String name) {
    if (name == null || name.isBlank()) {
      throw new RejectedException("template_name", "a template name must not be empty");
    }
    for (Rule rule : RULES) {
      if (rule.fires().test(name)) {
        throw new RejectedException(
            "template_name",
            "`" + Texts.bounded(name) + "` is not a template name: " + rule.explanation());
      }
    }
  }

  /**
   * The check a name-shape rule cannot make.
   *
   * <p>After following whatever the filesystem has there, is the answer still inside the root? A
   * symlink is what this exists for — it passes every rule above and still points out.
   *
   * <p>The existence check is explicit, so a missing template is named as one rather than falling
   * through to a confusing read error. And the containment test is STRUCTURAL rather than a string
   * prefix compare, which a sibling directory named {@code root-evil} would beat.
   */
  private Path contained(Path directory) {
    Path root;
    Path real;
    try {
      root = Paths.get(path).toRealPath();
      real = directory.toRealPath();
    } catch (IOException | RuntimeException error) {
      throw new RejectedException(
          "template_not_found", "no template by that name", String.valueOf(error.getMessage()));
    }

    if (real.equals(root) || isUnder(real, root)) {
      return real;
    }
    throw new RejectedException(
        "template_escapes_root", "the template resolves outside the template root");
  }

  private static boolean isUnder(Path real, Path root) {
    for (Path parent = real.getParent(); parent != null; parent = parent.getParent()) {
      if (parent.equals(root)) {
        return true;
      }
    }
    return false;
  }

  private static String read(Path file) {
    try {
      return Files.readString(file, StandardCharsets.UTF_8);
    } catch (IOException | RuntimeException error) {
      throw new RejectedException(
          "template_unreadable",
          "the template could not be read",
          String.valueOf(error.getMessage()));
    }
  }

  private static String optional(Path file) {
    return Files.isRegularFile(file, LinkOption.NOFOLLOW_LINKS) || Files.isRegularFile(file)
        ? read(file)
        : null;
  }

  private record Rule(Predicate<String> fires, String explanation) {}
}

File: sdk/dotnet/Shojiku/TemplateRoot.cs
// Resolving a template NAME to the sources behind it.
//
// A name is an identifier, never a path. A bundle format will take this lookup
// over later, so nothing outside this class may assume a directory is how names
// resolve — callers ask for "receipt_ja" and get sources back.
//
// THE REJECTION RULES ARE THE UNION ACROSS PLATFORMS, NOT THE HOST'S. Windows is
// a first-class target (it is what this SDK's market runs on), so a backslash is
// a separator, `C:name` is drive-relative, `\\host\share` is a UNC path and
// CON/NUL are reserved devices — every one of them refused on EVERY platform. A
// template name that is valid on one machine is valid on all of them, which is
// the only way the same application deploys to both.

using System.Text.RegularExpressions;

namespace Shojiku;

/// <summary>
/// A refused name or an unreadable template.
/// </summary>
/// <remarks>
/// Rejection is an exception INSIDE this class and a failed result outside it —
/// a hostile template name is a fact about the request, not a bug in the calling
/// program.
/// </remarks>
internal sealed class RejectedException : Exception
{
    internal RejectedException(string kind, string message, string? causeMessage = null)
        : base(message)
    {
        Kind = kind;
        CauseMessage = causeMessage;
    }

    internal string Kind { get; }

    internal string? CauseMessage { get; }
}

/// <summary>One configured root, and the only thing that turns names into sources.</summary>
internal sealed partial class TemplateRoot
{
    internal const string TemplateFile = "templates.yml";
    internal const string DefinitionsFile = "definitions.yml";

    /// <summary>
    /// Reserved DOS device names.
    /// </summary>
    /// <remarks>Windows resolves these no matter what directory you are in and no matter what extension you append.</remarks>
    private static readonly HashSet<string> Devices = BuildDevices();

    private readonly string path;

    internal TemplateRoot(string path) => this.path = path;

    /// <summary>The configured root directory.</summary>
    internal string Path => path;

    /// <summary>Resolve <paramref name="name"/>, or throw <see cref="RejectedException"/> naming why it will not.</summary>
    internal Sources Resolve(string name)
    {
        Reject(name);
        var real = Contained(System.IO.Path.Combine(path, name));
        return new Sources(
            Read(System.IO.Path.Combine(real, TemplateFile)),
            Optional(System.IO.Path.Combine(real, DefinitionsFile)),
            real);
    }

    // A name is ONE segment. Refusing both separators outright subsumes
    // traversal, absolute paths and nested lookups in a single rule — the
    // simplest thing six other SDKs can mirror without drifting.
    [GeneratedRegex(@"[/\\]")]
    private static partial Regex Separators();

    [GeneratedRegex(@"\A[A-Za-z]:")]
    private static partial Regex DriveRelative();

    [GeneratedRegex(@"[.\s]+\z")]
    private static partial Regex TrailingDotsAndSpaces();

    private static HashSet<string> BuildDevices()
    {
        var devices = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "CON", "PRN", "AUX", "NUL" };
        for (var n = 1; n <= 9; n++)
        {
            devices.Add($"COM{n}");
            devices.Add($"LPT{n}");
        }

        return devices;
    }

    private static bool IsSeparator(string name) => Separators().IsMatch(name);

    private static bool IsControl(string name)
    {
        foreach (var character in name)
        {
            if (char.IsControl(character))
            {
                return true;
            }
        }

        return false;
    }

    private static bool IsDriveRelative(string name) => DriveRelative().IsMatch(name);

    /// <summary>
    /// Trailing dots and spaces are STRIPPED by Windows before it resolves a name.
    /// </summary>
    /// <remarks>
    /// So <c>CON.</c> and <c>"CON "</c> are the CON device just as <c>CON</c> is.
    /// Without that strip they slip past this rule and are refused later, by
    /// containment — still refused, but with a message about a missing template
    /// rather than about a reserved name.
    /// </remarks>
    private static bool IsDevice(string name)
    {
        var stem = TrailingDotsAndSpaces().Replace(name.Split('.')[0], string.Empty);
        return Devices.Contains(stem);
    }

    private static void Reject(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            throw new RejectedException("template_name", "a template name must not be empty");
        }

        // Each rule, and what a caller is told when it fires.
        (Func<string, bool> Fires, string Explanation)[] rules =
        [
            (IsSeparator,
                "a name is one segment, so `/` and `\\` are never part of it "
                + "(which is also what makes `..` traversal impossible)"),
            (IsControl, "it contains a control character"),
            (IsDriveRelative,
                "it is drive-relative, which Windows resolves against that drive's current directory"),
            (IsDevice, "it is a reserved device name on Windows"),
        ];

        foreach (var (fires, explanation) in rules)
        {
            if (fires(name))
            {
                throw new RejectedException(
                    "template_name",
                    $"`{Text.Bounded(name)}` is not a template name: {explanation}");
            }
        }
    }

    private static string Read(string file)
    {
        try
        {
            return File.ReadAllText(file, System.Text.Encoding.UTF8);
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException or ArgumentException or NotSupportedException)
        {
            throw new RejectedException("template_unreadable", "the template could not be read", error.Message);
        }
    }

    private static string? Optional(string file) => File.Exists(file) ? Read(file) : null;

    /// <summary>
    /// The check a name-shape rule cannot make.
    /// </summary>
    /// <remarks>
    /// After following whatever the filesystem has there, is the answer still
    /// inside the root? A symlink is what this exists for — it passes every rule
    /// above and still points out.
    /// <para>
    /// The existence check is explicit: canonicalization here does NOT fail for
    /// a path that is not there, so a missing template would canonicalize
    /// happily and fall through to a confusing read error. And the containment
    /// test is STRUCTURAL rather than a string prefix compare, which a sibling
    /// directory named <c>root-evil</c> would beat.
    /// </para>
    /// </remarks>
    private string Contained(string directory)
    {
        string root;
        string real;
        try
        {
            if (!Directory.Exists(path) || !Directory.Exists(directory))
            {
                throw new DirectoryNotFoundException($"no directory at {directory}");
            }

            root = Canonical(path);
            real = Canonical(directory);
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException or ArgumentException or NotSupportedException)
        {
            throw new RejectedException("template_not_found", "no template by that name", error.Message);
        }

        if (string.Equals(real, root, StringComparison.Ordinal) || IsUnder(real, root))
        {
            return real;
        }

        throw new RejectedException("template_escapes_root", "the template resolves outside the template root");
    }

    /// <summary>Absolute, symlinks followed, and with NO trailing separator.</summary>
    /// <remarks>
    /// The trim is what keeps this SDK's canonical form the same SHAPE as the other
    /// six, whose canonicalizers (<c>realpath</c> and its equivalents) drop a trailing
    /// separator as part of what they do. .NET's do not — <c>DirectoryInfo.FullName</c>
    /// and <c>Path.GetFullPath</c> both preserve it — so a root configured as
    /// <c>templates/</c> canonicalized to <c>/app/templates/</c> while every parent
    /// <see cref="IsUnder"/> walks canonicalizes without one. No comparison could match,
    /// and a perfectly good root was reported as <c>template_escapes_root</c>. Relative
    /// versus absolute never mattered; the separator did.
    /// <para>
    /// <c>Path.TrimEndingDirectorySeparator</c> rather than a hand-rolled <c>TrimEnd</c>:
    /// it deliberately leaves a ROOT path alone, so <c>/</c> stays <c>/</c> and
    /// <c>C:\</c> stays <c>C:\</c> instead of becoming an empty string and a
    /// drive-relative <c>C:</c>.
    /// </para>
    /// <para>
    /// Normalizing HERE, not at the comparison: containment stays the structural walk
    /// below. Making the root comparable by string-prefix instead would admit a sibling
    /// directory named <c>&lt;root&gt;-evil</c>, which is exactly what that walk exists
    /// to refuse.
    /// </para>
    /// </remarks>
    private static string Canonical(string directory)
    {
        var info = new DirectoryInfo(directory);
        var resolved = info.ResolveLinkTarget(returnFinalTarget: true)?.FullName ?? info.FullName;
        return System.IO.Path.TrimEndingDirectorySeparator(System.IO.Path.GetFullPath(resolved));
    }

    private static bool IsUnder(string real, string root)
    {
        var parent = Directory.GetParent(real);
        while (parent is not null)
        {
            if (string.Equals(parent.FullName, root, StringComparison.Ordinal))
            {
                return true;
            }

            parent = parent.Parent;
        }

        return false;
    }
}
